openfe/protocols/openmm_utils/charge_generation.py

```diff
@@ -0,0 +1,415 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+"""
+Reusable utilities for assigning partial charges to ChemicalComponents.
+"""
+import copy
+from typing import Union, Optional, Literal, Callable
+import sys
+import numpy as np
+from openff.units import unit
+from openff.toolkit import Molecule as OFFMol
+from openff.toolkit.utils.base_wrapper import ToolkitWrapper
+from openff.toolkit.utils.toolkits import (
+    AmberToolsToolkitWrapper,
+    OpenEyeToolkitWrapper,
+    RDKitToolkitWrapper
+)
+from openff.toolkit.utils.toolkit_registry import ToolkitRegistry
+
+try:
+    import openeye
+except ImportError:
+    HAS_OPENEYE = False
+else:
+    HAS_OPENEYE = True
+
+try:
+    from openff.toolkit.utils.toolkit_registry import (
+        toolkit_registry_manager,
+    )
+except ImportError:
+    # toolkit_registry_manager was made non private in 0.14.4
+    from openff.toolkit.utils.toolkit_registry import (
+        _toolkit_registry_manager as toolkit_registry_manager
+    )
+
+
+try:
+    from openff.toolkit.utils.nagl_wrapper import NAGLToolkitWrapper
+    from openff.nagl_models.openff_nagl_models import (
+        get_models_by_type, validate_nagl_model_path
+    )
+except ImportError:
+    HAS_NAGL = False
+else:
+    HAS_NAGL = True
+
+
+try:
+    from espaloma_charge.openff_wrapper import EspalomaChargeToolkitWrapper
+except ImportError:
+    HAS_ESPALOMA = False
+else:
+    HAS_ESPALOMA = True
+
+
+# Dictionary of lists for the various backend options we allow.
+# Note: can't create the classes ahead of time in case we end
+# up with a case where the tool is not available, e.g. if OpenEye tk
+# is not installed.
+BACKEND_OPTIONS: dict[str, list[ToolkitWrapper]] = {
+    "ambertools": [RDKitToolkitWrapper, AmberToolsToolkitWrapper],
+    "openeye": [OpenEyeToolkitWrapper],
+    "rdkit": [RDKitToolkitWrapper],
+}
+
+
+def assign_offmol_espaloma_charges(
```

Comment:
Espaloma charge hasn't been updated in a while, @mikemhenry do you have any insights on the status? I was under the understanding that a new model was coming out soon?

Reply:
If the status is ambiguous then we should just remove this for now.

Reply:
1) I will make sure espaloma_charge is updated to use newer model and better fetching
2) I will update espaloma_charge on conda-forge

Maybe you could add a warning that using espaloma isn't well tested until we spruce things up. I think it is better to keep it and make it better than remove it and add it back in when it is working better